### PR TITLE
Add native histogram bucket limit

### PR DESCRIPTION
**Pull request by krajorama**
#### What this PR does

Memory protection from ingesting unreasonably huge samples. Classic sample is fixed size with a timestamp and value, but native histogram may contain 2^31 buckets.

Out of scope: validate histogram makes sense, only check valid histogram.

#### Which issue(s) this PR fixes or relates to

Relates to: #4542 

#### Checklist

- [x] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by zenador** [comment]
LGTM but someone else should probably take a look as well, and should changelog be updated?

File: pkg/distributor/distributor_test.go [hunk 1]
```diff
@@ -1519,13 +1524,24 @@ func TestDistributor_Push_HistogramValidation(t *testing.T) {
 			errMsg: "received a sample whose timestamp is too far in the future",
 			errID:  globalerror.SampleTooFarInFuture,
 		},
+		"buckets at limit": {
```
Comment (charleskorn) [inline pkg/distributor/distributor_test.go:1527]: Should we add some variations of the test case where all of the buckets are negative or positive, or a mixture of positive and negative? (not sure whether this is best here or in the more detailed tests below in `validate_test.go`)

Comment (krajorama) [reply pkg/distributor/distributor_test.go:1527]: I'll add in validate_test.go

**Review by osg-grafana — APPROVED** [review]

**Comment by krajorama** [comment]
I need to either analyze existing histograms to verify my idea of the default value for the limit or just go with 0 for now.

**Review by charleskorn — APPROVED** [review]

**Pull request merged**